src/Npgsql/ConnectorPool.Multiplexing.cs

```diff
@@ -0,0 +1,459 @@
+
+        const int WriteCoalescineDelayAdaptivityUs = 10;
+
+        /// <summary>
+        /// A pool-wide type mapper used when multiplexing. This is necessary because binding parameters
+        /// to their type handlers happens *before* the command is enqueued for execution, so there's no
+        /// connector yet at that stage.
+        /// </summary>
+        internal ConnectorTypeMapper? MultiplexingTypeMapper { get; private set; }
+
+        /// <summary>
+        /// When multiplexing is enabled, determines the maximum amount of time to wait for further
+        /// commands before flushing to the network. In ticks (100ns), 0 disables waiting.
+        /// This is in 100ns ticks, not <see cref="Stopwatch"/> ticks whose meaning vary across platforms.
+        /// </summary>
+        readonly long _writeCoalescingDelayTicks;
+
+        /// <summary>
+        /// When multiplexing is enabled, determines the maximum number of outgoing bytes to buffer before
+        /// flushing to the network.
+        /// </summary>
+        readonly int _writeCoalescingBufferThresholdBytes;
+
+        internal long NumCommandsSent;
+        internal long NumBatches;
+        long _ticksWritten;
+        long _waitsForFurtherCommands;
+        // long _bytesFlushed;
+
+        /// <summary>
+        /// Called exactly once per multiplexing pool, when the first connection is opened, with two goals:
+        /// 1. Load types and bind the pool-wide type mapper (necessary for binding parameters)
+        /// 2. Cause any connection exceptions (e.g. bad username) to be thrown from NpgsqlConnection.Open
+        /// </summary>
+        internal async Task BootstrapMultiplexing(NpgsqlConnection conn, NpgsqlTimeout timeout, bool async, CancellationToken cancellationToken = default)
+        {
+            Debug.Assert(_multiplexing);
+
+            var connector =
+                await conn.StartBindingScope(ConnectorBindingScope.Connection, timeout, async, cancellationToken);
+            using var _ = Defer(() => conn.EndBindingScope(ConnectorBindingScope.Connection));
+
+            // Somewhat hacky. Extract the connector's type mapper as our pool-wide mapper,
+            // and have the connector rebind to ensure it has a different instance.
+            // The latter isn't strictly necessary (type mappers should always be usable
+            // concurrently) but just in case.
+            MultiplexingTypeMapper = connector.TypeMapper;
+            connector.RebindTypeMapper();
+
+            IsBootstrapped = true;
+        }
+
+        async Task MultiplexingWriteLoopWrapper()
+        {
+            try
+            {
+                await MultiplexingWriteLoop();
+            }
+            catch (Exception e)
+            {
+                Log.Error("Exception in multiplexing write loop, this is an Npgsql bug, please file an issue.", e);
+            }
+        }
+
+        async Task MultiplexingWriteLoop()
```

Comment:
Very correct, and it's somewhat central to the design - there are various assumptions that this code will never run concurrently, so no locking/sync needed. At some scale this could become a problem, especially if encoding the wire data is somehow very expensive (e.g. huge UTF8 strings?).

Long-term, I definitely want to move encoding out of here to before the command channel, leaving this code to just get buffers out of the channel and write them - nothing more. However, this would depend on:

* Usage of gather I/O to collect buffers from the different commands from the channel, and write them into a single packet in one go (hence https://github.com/dotnet/runtime/issues/36951). This doesn't worry me too much (it's not amazing but is OK). However:
* We'd need a different model for auto-prepared statements, since before we enqueue a command to the channel we have no connector, so no idea if the SQL is prepared or not... 

Stepping back, multiplexing won't immediately be the best thing for every scenario - there are probably some cases where it's going to be worse. Or it may reduce throughput a bit, but at the same time considerably reduce the number of physical connections that are needed (thus reducing strain on PG). At the beginning I had hoped this would be a "just turn it on and rock" kind of feature, but it's (unfortunately) more complicated than that.

Opportunities to improve in the future :)